File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3076_shortest-uncommon-substring-in-an-array/solutions/cpp/solution.cpp
#include <vector>
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
public:
    vector<string> shortestSubstrings(vector<string>& arr) {
        int n = arr.size();
        vector<string> answer(n, "");
        
        for (int i = 0; i < n; ++i) {
            unordered_set<string> otherSubstrings;
            string current = arr[i];
            
            // Collect all substrings from other strings
            for (int j = 0; j < n; ++j) {
                if (i == j) continue;
                string other = arr[j];
                int len = other.length();
                
                for (int k = 0; k < len; ++k) {
                    for (int l = 1; l <= len - k; ++l) {
                        otherSubstrings.insert(other.substr(k, l));
                    }
                }
            }
            
            // Find the shortest and lexicographically smallest substring not in otherSubstrings
            int len = current.length();
            string best = "";
            
            for (int start = 0; start < len; ++start) {
                for (int length = 1; length <= len - start; ++length) {
                    string sub = current.substr(start, length);
                    if (otherSubstrings.find(sub) == otherSubstrings.end()) {
                        if (best.empty() || length < best.length() || (length == best.length() && sub < best)) {
                            best = sub;
                        }
                    }
                }
                // Early exit if we've found the smallest possible unique substring
                if (!best.empty() && best.length() == 1) break;
            }
            
            answer[i] = best;
        }
        
        return answer;
    }
};

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3076_shortest-uncommon-substring-in-an-array/solutions/python3/solution.py
class Solution:
    def shortestSubstrings(self, arr: list[str]) -> list[str]:
        from itertools import combinations
        
        def all_substrings(s):
            n = len(s)
            return [s[i:j] for i in range(n) for j in range(i + 1, n + 1)]
        
        n = len(arr)
        answer = [''] * n
        
        # Create a dictionary to hold all substrings of each string
        substrings = {i: all_substrings(arr[i]) for i in range(n)}
        
        # Attempt to find the shortest unique substring
        for i in range(n):
            found = False
            # Sort substrings by length and then lexicographically
            sorted_subs = sorted(set(substrings[i]), key=lambda x: (len(x), x))
            for sub in sorted_subs:
                is_unique = True
                # Check if this substring appears in any other string
                for j in range(n):
                    if i != j and sub in arr[j]:
                        is_unique = False
                        break
                if is_unique:
                    answer[i] = sub
                    found = True
                    break
            if not found:
                answer[i] = ''
        
        return answer
